File: sdk/python/src/protomcp/transport.py
import socket
import struct
import sys
import os

# Add gen directory to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..', 'gen'))
import protomcp_pb2 as pb

class Transport:
    def __init__(self, socket_path: str):
        self._socket_path = socket_path
        self._sock: socket.socket | None = None

    def connect(self):
        self._sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        self._sock.connect(self._socket_path)

    def send(self, envelope: pb.Envelope):
        data = envelope.SerializeToString()
        length = struct.pack(">I", len(data))
        self._sock.sendall(length + data)

    def send_chunked(self, request_id: str, field_name: str, data: bytes,
                     chunk_size: int = 65536):
        """Send a large field as StreamHeader + StreamChunk messages."""
        header = pb.Envelope(
            request_id=request_id,
            stream_header=pb.StreamHeader(
                field_name=field_name,
                total_size=len(data),
                chunk_size=chunk_size,
            ),
        )
        self.send(header)

        offset = 0
        while offset < len(data):
            end = min(offset + chunk_size, len(data))
            is_final = (end >= len(data))
            chunk_env = pb.Envelope(
                request_id=request_id,
                stream_chunk=pb.StreamChunk(
                    data=data[offset:end],
                    final=is_final,
                ),
            )
            self.send(chunk_env)
            offset = end

    def send_raw(self, request_id: str, field_name: str, data: bytes):
        """Send a large field as a RawHeader + raw bytes (no protobuf wrapping on payload)."""
        compression = ""
        uncompressed_size = 0
        threshold = int(os.environ.get("PROTOMCP_COMPRESS_THRESHOLD", "65536"))
        if len(data) > threshold:
            import zstandard
            compressor = zstandard.ZstdCompressor()
            uncompressed_size = len(data)
            data = compressor.compress(data)
            compression = "zstd"
        header = pb.Envelope(
            raw_header=pb.RawHeader(
                request_id=request_id,
                field_name=field_name,
                size=len(data),
                compression=compression,
                uncompressed_size=uncompressed_size,
            ),
        )
        # Send the protobuf header normally
        header_bytes = header.SerializeToString()
        length = struct.pack(">I", len(header_bytes))
        # Send header + raw payload in one sendall to minimize syscalls
        self._sock.sendall(length + header_bytes + data)

    def recv(self) -> pb.Envelope:
        length_bytes = self._recv_exactly(4)
        length = struct.unpack(">I", length_bytes)[0]
        data = self._recv_exactly(length)
        env = pb.Envelope()
        env.ParseFromString(data)
        return env

    def close(self):
        if self._sock:
            self._sock.close()

    def _recv_exactly(self, n: int) -> bytes:
        buf = bytearray()
        while len(buf) < n:
            chunk = self._sock.recv(n - len(buf))
            if not chunk:
                raise ConnectionError("socket closed")
            buf.extend(chunk)
        return bytes(buf)

File: sdk/go/protomcp/transport.go
package protomcp

import (
	"encoding/binary"
	"fmt"
	"io"
	"net"
	"os"
	"strconv"

	"github.com/klauspost/compress/zstd"
	pb "github.com/msilverblatt/protomcp/gen/proto/protomcp"
	"google.golang.org/protobuf/proto"
)

type Transport struct {
	socketPath string
	conn       net.Conn
}

func NewTransport(socketPath string) *Transport {
	return &Transport{socketPath: socketPath}
}

func (t *Transport) Connect() error {
	conn, err := net.Dial("unix", t.socketPath)
	if err != nil {
		return fmt.Errorf("connect to socket: %w", err)
	}
	t.conn = conn
	return nil
}

func (t *Transport) Send(env *pb.Envelope) error {
	data, err := proto.Marshal(env)
	if err != nil {
		return fmt.Errorf("marshal envelope: %w", err)
	}
	length := make([]byte, 4)
	binary.BigEndian.PutUint32(length, uint32(len(data)))
	if _, err := t.conn.Write(length); err != nil {
		return err
	}
	_, err = t.conn.Write(data)
	return err
}

// SendRaw sends a RawHeader envelope followed by raw payload bytes.
// This avoids protobuf serialization overhead for large payloads.
// If the payload exceeds PROTOMCP_COMPRESS_THRESHOLD (default 64KB),
// it is zstd-compressed before sending.
func (t *Transport) SendRaw(requestID, fieldName string, data []byte) error {
	compression := ""
	var uncompressedSize uint64
	threshold := 65536
	if v := os.Getenv("PROTOMCP_COMPRESS_THRESHOLD"); v != "" {
		if n, err := strconv.Atoi(v); err == nil {
			threshold = n
		}
	}
	if len(data) > threshold {
		encoder, err := zstd.NewWriter(nil)
		if err != nil {
			return fmt.Errorf("create zstd encoder: %w", err)
		}
		compressed := encoder.EncodeAll(data, make([]byte, 0, len(data)))
		encoder.Close()
		uncompressedSize = uint64(len(data))
		data = compressed
		compression = "zstd"
	}

	header := &pb.Envelope{
		Msg: &pb.Envelope_RawHeader{
			RawHeader: &pb.RawHeader{
				RequestId:        requestID,
				FieldName:        fieldName,
				Size:             uint64(len(data)),
				Compression:      compression,
				UncompressedSize: uncompressedSize,
			},
		},
	}
	headerBytes, err := proto.Marshal(header)
	if err != nil {
		return fmt.Errorf("marshal raw header: %w", err)
	}
	length := make([]byte, 4)
	binary.BigEndian.PutUint32(length, uint32(len(headerBytes)))

	// Write length-prefixed header + raw payload
	buf := make([]byte, 0, 4+len(headerBytes)+len(data))
	buf = append(buf, length...)
	buf = append(buf, headerBytes...)
	buf = append(buf, data...)
	_, err = t.conn.Write(buf)
	return err
}

func (t *Transport) Recv() (*pb.Envelope, error) {
	lengthBuf := make([]byte, 4)
	if _, err := io.ReadFull(t.conn, lengthBuf); err != nil {
		return nil, err
	}
	length := binary.BigEndian.Uint32(lengthBuf)
	data := make([]byte, length)
	if _, err := io.ReadFull(t.conn, data); err != nil {
		return nil, err
	}
	env := &pb.Envelope{}
	if err := proto.Unmarshal(data, env); err != nil {
		return nil, fmt.Errorf("unmarshal envelope: %w", err)
	}
	return env, nil
}

func (t *Transport) Close() {
	if t.conn != nil {
		t.conn.Close()
	}
}
